transport/pool.go

```diff
@@ -0,0 +1,252 @@
+			// Time interval of refilling is adjusted based on what caused it.
+			if p.closeSinceRefill {
+				p.onClose()
+			} else {
+				p.currentDelay = minFillBackoff
+			}
+
+			ticker.Reset(p.currentDelay)
+			refilling = true
+		} else if !p.needsFilling() {
+			ticker.Stop()
+		}
+	}
+}
+
+// Actual connection closer.
+func (p *PoolRefiller) closeConn(idx uint16) {
+	conn := p.connPool.loadConn(idx)
+	if conn == nil {
+		return
+	}
+	conn.Close()
+	p.connPool.storeConn(idx, nil)
+	p.activeShards--
+	p.closeSinceRefill = true
+}
+
+// Time interval of refilling is adjusted here.
+func (p *PoolRefiller) onClose() {
+	p.closeSinceRefill = false
+	if maxFillBackoff > p.currentDelay*fillBackoffMultiplier {
+		p.currentDelay *= fillBackoffMultiplier
+	} else {
+		p.currentDelay = maxFillBackoff
+	}
+}
+
+// Check whether ConnPool is full of connections,
+// will be more complicated in the future.
+func (p *PoolRefiller) needsFilling() bool {
+	return p.activeShards < p.connPool.nrShards
+}
+
+// Opens first connection of ConnPool and fills read values from scyllaSupported.
+func (p *PoolRefiller) firstConnection() {
+	// First connection is needed to be done a bit differently,
+	// without opening it to a specific shard in order to read frame.ScyllaSupported.
+	conn, err := OpenConn(p.address, nil, p.connConfig)
+	if err != nil {
+		return
+	}
+	p.connOptions, err = getScyllaSupported(conn)
+	if err != nil {
+		return
+	}
+
+	p.connPool.nrShards = p.connOptions.NrShards
+	p.connPool.conns = make([]atomic.Value, p.connPool.nrShards)
+	p.connPool.storeConn(p.connOptions.Shard, conn)
+	p.connPool.msbIgnore = p.connOptions.MsbIgnore
+	p.connPool.alertChan = make(chan uint16, p.connPool.nrShards)
+	p.activeShards = 1
+}
+
+// Refills ConnPool.
+func (p *PoolRefiller) fill() {
+	if p.activeShards == 0 && p.connPool.nrShards == 0 {
+		p.firstConnection()
+	}
+
+	n := p.connOptions.NrShards
+	for i := uint16(0); i < n && p.activeShards < p.connPool.nrShards; i++ {
+		if p.connPool.loadConn(i) != nil {
+			continue
+		}
+
+		shard := ShardInfo{
+			Shard:     i,
+			NrShards:  n,
+			MsbIgnore: p.connPool.msbIgnore,
+		}
+
+		conn, err := OpenShardConn(p.address, shard, p.connConfig)
+		if err == nil {
+			p.connPool.storeConn(shard.Shard, conn)
+			p.activeShards++
+		} else {
+			p.closeSinceRefill = true
+		}
+	}
+}
+
+// Sends request for frame.Supported struct and returns either frame.ScyllaSupported or error.
+func getScyllaSupported(conn *Conn) (frame.ScyllaSupported, error) {
+	res, err := conn.sendRequest(&Options{}, false, false)
+	if err != nil {
+		return frame.ScyllaSupported{}, err
+	}
+	supp, ok := res.(*Supported)
+	if !ok {
+		return frame.ScyllaSupported{}, errors.New("couldn't cast interface to struct")
+	}
+
+	return *supp.ParseScyllaSupported(), nil
+}
+
+// Initiates ConnPool along with its PoolRefiller.
+func InitNodeConnPool(addr string, connConfig ConnConfig) *ConnPool { // nolint:unused // This will be used.
```

Comment:
Docs on exported functions must start with the function name there is a linter for that.

Reply:
Better add a dummy use.